File: app/src/main/java/com/kimikevin/eatsplorer/model/entity/NearbySearchRequest.java
package com.kimikevin.eatsplorer.model.entity;

import com.google.gson.annotations.SerializedName;

public class NearbySearchRequest {
    @SerializedName("includedTypes")
    private String[] includedTypes;
    @SerializedName("maxResultCount")
    private int maxResultCount;
    private LocationRestriction locationRestriction;

    public NearbySearchRequest(double latitude, double longitude, double radiusMeters) {
        this.includedTypes = new String[] {"restaurant"};
        this.maxResultCount = 20;
        this.locationRestriction = new LocationRestriction(
                new Circle(new Center(latitude, longitude), radiusMeters)
        );
    }

    private record LocationRestriction(@SerializedName("circle") Circle circle) {
    }

    private record Circle(@SerializedName("center") Center center,
                          @SerializedName("radius") double radius) {
    }

    private record Center(@SerializedName("latitude") double latitude,
                          @SerializedName("longitude") double longitude) {
    }
}

File: app/src/main/java/com/kimikevin/eatsplorer/model/entity/NearbySearchRequest.kt
package com.kimikevin.eatsplorer.model.entity

import com.google.gson.annotations.SerializedName

data class NearbySearchRequest(
    @SerializedName("includedTypes")
    val includedTypes: Array<String> = arrayOf("restaurant"),
    @SerializedName("maxResultCount")
    val maxResultCount: Int = 20,
    val locationRestriction: LocationRestriction
) {
    constructor(latitude: Double, longitude: Double, radiusMeters: Double) : this(
        locationRestriction = LocationRestriction(
            circle = Circle(
                center = Center(latitude, longitude),
                radius = radiusMeters
            )
        )
    )

    data class LocationRestriction(
        @SerializedName("circle")
        val circle: Circle
    )

    data class Circle(
        @SerializedName("center")
        val center: Center,
        @SerializedName("radius")
        val radius: Double
    )

    data class Center(
        val latitude: Double,
        val longitude: Double
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as NearbySearchRequest
        if (!includedTypes.contentEquals(other.includedTypes)) return false
        return true
    }

    override fun hashCode(): Int {
        return includedTypes.contentHashCode()
    }
}
